File: backend/app/schemas/ats.py
from pydantic import BaseModel, field_validator
from typing import List, Optional


class ATSAnalyzeRequest(BaseModel):
    cv_id: str
    job_description: str


class DownloadRequest(BaseModel):
    cv_id: str


class FetchJobRequest(BaseModel):
    url: str


class ScoreBreakdown(BaseModel):
    label: str
    score: int
    icon: str = "key"

    @field_validator("score", mode="before")
    @classmethod
    def coerce_score(cls, v):
        return int(round(float(v)))


class DiffChange(BaseModel):
    section: str = "General"
    before: str = ""
    after: str = ""


class Keyword(BaseModel):
    word: str
    present: bool = False
    priority: str = "medium"  # high, medium, low


class ComparisonItem(BaseModel):
    requirement: str = ""
    cv_value: str = ""
    status: str = "missing"  # match | missing | partial

    @field_validator("status", mode="before")
    @classmethod
    def coerce_status(cls, v):
        v = str(v).lower().strip()
        return v if v in ("match", "missing", "partial") else "missing"


class ATSAnalysisResult(BaseModel):
    overall_score: int
    breakdown: List[ScoreBreakdown] = []
    missing_keywords: List[str] = []
    present_keywords: List[str] = []
    suggestions: List[str] = []
    diff_changes: List[DiffChange] = []
    keyword_match_pct: int = 0
    comparison: List[ComparisonItem] = []

    @field_validator("overall_score", "keyword_match_pct", mode="before")
    @classmethod
    def coerce_int(cls, v):
        return int(round(float(v)))


class ApplyChangesRequest(BaseModel):
    cv_id: str
    accepted_changes: List[int]
    added_keywords: List[str]


class ATSDownloadRequest(BaseModel):
    cv_id: str
    diff_changes: List[DiffChange]      # full list from the analysis result
    accepted_changes: List[int]         # indices of changes accepted by the user
    added_keywords: List[str]           # keywords the user clicked to add


class JobPostingData(BaseModel):
    title: Optional[str] = None
    company: Optional[str] = None
    location: Optional[str] = None
    description: str
    requirements: List[str] = []

File: frontend/lib/api/ats.ts
import { request, download, ApiError } from "./client";
import { getIdToken } from "@/lib/firebase";

export interface ScoreBreakdown {
  label: string;
  score: number;
  icon: string;
}

export interface DiffChange {
  section: string;
  before: string;
  after: string;
}

export interface ComparisonItem {
  requirement: string;
  cv_value: string;
  status: "match" | "missing" | "partial";
}

export interface ATSAnalysisResult {
  overall_score: number;
  breakdown: ScoreBreakdown[];
  missing_keywords: string[];
  present_keywords: string[];
  suggestions: string[];
  diff_changes: DiffChange[];
  keyword_match_pct: number;
  comparison: ComparisonItem[];
}

export interface JobPostingData {
  title?: string;
  company?: string;
  location?: string;
  description: string;
  requirements: string[];
}

export const atsApi = {
  /**
   * Calls the Next.js Route Handler at /api/ats-analyze (not the rewrites proxy)
   * to avoid ECONNRESET / socket hang-up during long Gemini AI calls.
   */
  analyze: async (cvId: string, jobDescription: string): Promise<ATSAnalysisResult> => {
    const token = await getIdToken();
    const res = await fetch("/api/ats-analyze", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        ...(token ? { Authorization: `Bearer ${token}` } : {}),
      },
      body: JSON.stringify({ cv_id: cvId, job_description: jobDescription }),
    });
    const data = await res.json().catch(() => null);
    if (!res.ok) {
      const detail = data?.detail;
      const message =
        typeof detail === "string" ? detail
        : Array.isArray(detail) ? detail.map((e: { msg?: string }) => e?.msg ?? JSON.stringify(e)).join("; ")
        : detail ? JSON.stringify(detail)
        : "ATS analysis failed";
      throw new ApiError(message, res.status, data);
    }
    return data as ATSAnalysisResult;
  },

  applyChanges: (cvId: string, acceptedChanges: number[], addedKeywords: string[]) =>
    request<{ message: string; cv: unknown }>("/ats/apply-changes", {
      method: "POST",
      body: JSON.stringify({
        cv_id: cvId,
        accepted_changes: acceptedChanges,
        added_keywords: addedKeywords,
      }),
    }),

  fetchJob: (url: string) =>
    request<JobPostingData>("/ats/fetch-job", {
      method: "POST",
      body: JSON.stringify({ url }),
    }),

  downloadTailored: (cvId: string) =>
    download(`/ats/download-tailored?cv_id=${encodeURIComponent(cvId)}`),

  downloadOptimized: (
    cvId: string,
    diffChanges: DiffChange[],
    acceptedChanges: number[],
    addedKeywords: string[],
  ) =>
    download("/ats/download-optimized", {
      method: "POST",
      body: JSON.stringify({
        cv_id: cvId,
        diff_changes: diffChanges,
        accepted_changes: acceptedChanges,
        added_keywords: addedKeywords,
      }),
    }),
};
